xdsl/interpreters/shaped_array.py

```diff
@@ -0,0 +1,57 @@
+"""
+A helper structure to represent instances of type MemrefType, TensorType, VectorType, etc.
+in the interpreter.
+"""
+
+from itertools import accumulate
+from math import prod
+import operator
+from typing import Generic, TypeVar
+from attr import dataclass
+
+_T = TypeVar("_T")
+
+
+@dataclass
+class ShapedArray(Generic[_T]):
+    data: list[_T]
+    shape: list[int]
+
+    def __post__init__(self):
+        assert prod(self.shape) == len(self.data)
+
+    def offset(self, indices: tuple[int, ...]) -> int:
+        """
+        Returns the index of the element in `self.data` for a given tuple of indices
+        """
+        if len(indices) != len(self.shape):
+            raise ValueError(f"Invalid indices {indices} for shape {self.shape}")
```

Comment:
I think in general if we call Python functions from the interpreter we should expect non InterpreterError errors. Actually, now that I think about it, it only makes sese to use InterpreterError if it's an error with the interpreter infrastructure, and should be other specific errors otherwise.